### DISCUSSION TITLE
KiwiBuilder::loadMorphBin(std::istream& is) 의 역할이 궁금합니다.

### DISCUSSION THREAD

**Discussion by emiatej9** [bab2min/Kiwi#47] [Q&A]
안녕하세요.

`src/KiwiBuilder.cpp` 코드를 보던 중에 아래 부분의 역할이 궁금하여 질문드립니다.

https://github.com/bab2min/Kiwi/blob/fa1023dbf6e4c64a55a31a77fe8f371940130832/src/KiwiBuilder.cpp#L297-L305

`serializer::readMany` 와 `serializer::toKey("KIWI")` 그리고 `formMap`이 어떤 것을 의미하는지 알 수 있을까요?

`Key`와 관련해서는 아래 구조체도 존재하던데 `Key`가 어떤 기능을 하는지도 궁금합니다.

https://github.com/bab2min/Kiwi/blob/fa60f455fc1dd5b52b0318c50f0670a9db491c5d/src/serializer.hpp#L92-L113

감사합니다. 🙏

**Comment by bab2min** [comment]
`serializer::writeMany`는 객체를 직렬화하여 파일에 쓰는 기능, `serializer::readMany`는 파일로부터 직렬화된 객체를 읽어들여 역직렬화하는 기능을 수행합니다. 
```c++
vector<int> value1{ 0, 1, 2, 3, 4, 5 };
string value2 = "test";
ofstream some_output{ "output.bin", ios_base::binary };
serializer::writeMany(some_output, value1, value2); // some_output에 value1, value2가 저장됨


ifstream some_input{ "output.bin", ios_base::binary };
vector<int> value1;
string value2;
serializer::readMany(some_input, value1, value2);
// some_input으로부터 value1 = [0, 1, 2, 3, 4, 5], value2 = "test"를 읽어옴.
```
`serialize::Key`타입은 파일의 매직넘버를 부여하기 위한 기능을 수행합니다. `writeMany`에 쓰일 경우 파일에 해당 값을 그대로 쓰고, `readMany`에 쓰일 경우 파일로부터 값을 읽어온 뒤 Key와 일치하는지 테스트합니다. 같을 경우 파일을 계속 읽어오고, 틀린 경우 exception을 발생시킵니다.

즉 `serializer::readMany(is, serializer::toKey("KIWI"), forms, morphemes);`는 `is` 파일에서 먼저 4byte를 읽어와 `KIWI`와 일치하는지 테스트한뒤, 이게 성공하면 forms과 morphemes을 읽어오는 기능을 수행합니다.

`formMap`은 각각의 FormCond에 고유한 index값을 부여하고, 이를 탐색하기 위해 쓰이는 일종의 dictionary입니다. 이를 통해 임의의 FormCond가 formMap에 속한 경우 고유한 index값으로 변환할 수 있습니다.
